swarm/pss/pss_test.go

```diff
@@ -839,6 +838,94 @@ outer:
 
 }
 
+func TestDeduplication(t *testing.T) {
+	var err error
+
+	clients, err := setupNetwork(3)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	var addrsize = 32
+	var loaddrhex string
+	err = clients[0].Call(&loaddrhex, "pss_baseAddr")
+	if err != nil {
+		t.Fatalf("rpc get node 1 baseaddr fail: %v", err)
+	}
+	loaddrhex = loaddrhex[:2+(addrsize*2)]
+	var roaddrhex string
+	err = clients[1].Call(&roaddrhex, "pss_baseAddr")
+	if err != nil {
+		t.Fatalf("rpc get node 2 baseaddr fail: %v", err)
+	}
+	roaddrhex = roaddrhex[:2+(addrsize*2)]
+	var xoaddrhex string
+	err = clients[2].Call(&xoaddrhex, "pss_baseAddr")
+	if err != nil {
+		t.Fatalf("rpc get node 3 baseaddr fail: %v", err)
+	}
+	xoaddrhex = xoaddrhex[:2+(addrsize*2)]
+
+	log.Info("peer", "l", loaddrhex, "r", roaddrhex, "x", xoaddrhex)
+
+	var topic string
+	err = clients[0].Call(&topic, "pss_stringToTopic", "foo:42")
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	time.Sleep(time.Millisecond * 250)
+
+	// retrieve public key from pss instance
+	// set this public key reciprocally
+	var rpubkey string
+	err = clients[1].Call(&rpubkey, "pss_getPublicKey")
+	if err != nil {
+		t.Fatalf("rpc get receivenode pubkey fail: %v", err)
+	}
+
+	time.Sleep(time.Millisecond * 500) // replace with hive healthy code
+
+	rmsgC := make(chan APIMsg)
+	rctx, cancel := context.WithTimeout(context.Background(), time.Second*1)
+	defer cancel()
+	rsub, err := clients[1].Subscribe(rctx, "pss", rmsgC, "receive", topic)
+	log.Trace("rsub", "id", rsub)
+	defer rsub.Unsubscribe()
+
+	// store public key for recipient
+	// zero-length address means forward to all
+	// we have just two peers, they will be in proxbin, and will both receive
+	err = clients[0].Call(nil, "pss_setPeerPublicKey", rpubkey, topic, "0x")
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// send and verify delivery
+	rmsg := []byte("xyzzy")
+	err = clients[0].Call(nil, "pss_sendAsym", rpubkey, topic, hexutil.Encode(rmsg))
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	var receivedok bool
+OUTER:
+	for {
+		select {
+		case <-rmsgC:
+			if receivedok {
+				t.Fatalf("duplicate message received")
+			}
+			receivedok = true
+		case <-rctx.Done():
+			break OUTER
```

Comment:
Could you explain this please? Why do we go back to the top of the `for` loop on `rctx.Done()`? I see this also in `pss_test.go` and I think it is wrong.